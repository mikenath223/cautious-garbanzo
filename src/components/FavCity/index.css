.card {
  border-radius: 4px;
  align-self: flex-start;
  padding-inline: 1.2rem;
  padding-block: 1.7rem;
  background-color: #ffffff;
  display: flex;
  flex-direction: column;
  gap: 0.7rem;
  position: relative;
}

.card-item {
  display: flex;
  align-items: center;
  gap: 0.7rem;
}

.remove-btn {
  color: #ffffff;
  padding: unset !important;
  background-color: #ff0000;
  width: 22px;
  height: 21px;
  top: 10px;
  right: 10px;
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
}

.city-img {
  width: 24px;
  height: auto;
}

@media (max-width: 520px) {
  .card {
    margin-inline: auto;
    max-width: 80vw;
  }
}
