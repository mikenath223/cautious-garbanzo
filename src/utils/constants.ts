import { IWeatherInfo } from "./types"

export const initFavCities: string[] = [
  "Tokyo",
  "Delhi",
  "Shanghai",
  "Dhaka",
  "Sao Paulo",
  "Mexico City",
  "Cairo",
  "Beijing",
  "Mumbai",
  "Osaka",
  "Chongqing",
  "Karachi",
  "Kinshasa",
  "Lagos",
  "Istanbul"
]

export const defaultCity: IWeatherInfo = {
  "request": {
    "type": "City",
    "query": "New York, United States of America",
  },
  "location": {
    "name": "New York",
    "country": "United States of America",
    "region": "New York",
    "localtime": "2019-09-07 08:14",
  },
  "current": {
    "observation_time": "12:14 PM",
    "temperature": 13,
    "weather_code": 113,
    "weather_icons": [
      "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
    ],
    "weather_descriptions": [
      "Sunny"
    ],
    "wind_speed": 0,
    "wind_degree": 349,
    "wind_dir": "N",
    "pressure": 1010,
    "precip": 0,
    "humidity": 90,
    "cloudcover": 0,
    "feelslike": 13,
    "uv_index": 4,
    "visibility": 16
  }
}

export const defaultNote = {
  id: "01",
  text: "Welcome"
}
