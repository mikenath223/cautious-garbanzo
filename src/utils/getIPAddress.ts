const fetchIPAddress = async () => {
  try {
    const response = await fetch('https://ipapi.co/json')
    const data = await response.json()
    return data
  } catch (err) {
    console.error(err)
  }
}

export default fetchIPAddress
