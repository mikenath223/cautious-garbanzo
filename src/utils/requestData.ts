import { IWeatherInfo, PromiseFulfilledResult } from "./types"

export const getCityWeather = async (city: string, isCheckCache: boolean): Promise<IWeatherInfo> => {
  try {
    const currentDate = new Date()
    const localStorageKey = `${city.toLowerCase().trim()}-${currentDate.getHours()}-${currentDate.getDate()}-${currentDate.getMonth()}-${currentDate.getFullYear()}`
    const cachedCityData = localStorage.getItem(localStorageKey)
    if (cachedCityData && isCheckCache) {
      const weatherData = JSON.parse(cachedCityData)
      return weatherData
    }
    const url = `http://api.weatherstack.com/current?access_key=04c73e2c81cd37fd215de627927cdd44&query=${city}`;
    const response = await fetch(url)
    const result = await response.json()
    if (result.error) {
      return result?.error?.info
    }
    localStorage.setItem(localStorageKey, JSON.stringify(result))
    return result
  } catch (error) {
    const errMsg = `Failed to retrieve ${city.trim()} data!`
    throw new Error(errMsg)
  }
}

const sortCities = (cities: string[]) => cities.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))

export const getFavsCityWeather = async (cities: string[]) => {
  const fetchRequests = sortCities(cities).map(city => getCityWeather(city, true));

  try {
    const citiesInfo = await Promise.allSettled(fetchRequests)
    const data = citiesInfo.filter(data => data.status === "fulfilled").map(data => (data as PromiseFulfilledResult<IWeatherInfo>).value)
    return data
  } catch (_) {
    return []
  }
}