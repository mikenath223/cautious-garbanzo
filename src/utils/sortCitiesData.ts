import { IWeatherInfo } from "./types";

const sortCitiesData = (cities: IWeatherInfo[] | undefined) => cities?.sort((a, b) => a.location.name.toLowerCase().localeCompare(b.location.name.toLowerCase()))

export default sortCitiesData
