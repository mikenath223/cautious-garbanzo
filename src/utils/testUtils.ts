import { render } from '@testing-library/react';

export const withRender = (component: JSX.Element) =>
  render(component);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => { };
  localStorage.getItem = (key: string) => null;
}