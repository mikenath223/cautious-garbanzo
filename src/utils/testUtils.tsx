import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export const withRender = (component: JSX.Element) =>
  render(<MemoryRouter>{component}</MemoryRouter>);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => {};
  localStorage.getItem = (key: string) => null;
};
