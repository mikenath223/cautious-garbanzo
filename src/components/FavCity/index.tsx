import { Link } from "react-router-dom";
import { IWeatherInfo } from "../../utils/types";
import Button from "../Button";
import "./index.css";

interface IFavCityProps {
  city: IWeatherInfo;
  onRemoveCity: (city: IWeatherInfo) => void;
}

export default function FavCity({ onRemoveCity, city }: IFavCityProps) {
  const handleRemoveCity = (event: React.SyntheticEvent) => {
    event.stopPropagation();
    event.preventDefault();
    onRemoveCity(city);
  };

  const onSelectCity = () => {
    localStorage.setItem("selected-city", JSON.stringify(city));
  };

  return (
    <li onClick={() => onSelectCity()}>
      <Link to={"/details"} className="card box-shadow" tabIndex={0}>
        <Button
          className="remove-btn"
          type={"button"}
          tabIndex={0}
          aria-label={"Delete favorite city"}
          onClick={handleRemoveCity}
        >
          -
        </Button>
        <div className="card-item">
          <img src="/assets/icons/city.png" alt="" className="city-img" />
          <h3>{city?.name}</h3>
        </div>
        <div className="card-item">
          <img
            src="/assets/icons/temperature.png"
            alt=""
            className="city-img"
          />
          <h3>
            {city?.main?.temp}
            <sup>o</sup>C
          </h3>
        </div>
      </Link>
    </li>
  );
}
