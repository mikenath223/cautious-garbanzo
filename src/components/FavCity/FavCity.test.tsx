import { defaultCity } from "../../utils/constants";
import { withRender } from "../../utils/testUtils";
import FavCity from "./index";
import { MemoryRouter } from "react-router-dom";

describe("FavCity", function () {
  it("should match snapshot", () => {
    const { container } = withRender(
      <MemoryRouter>
        <FavCity city={defaultCity} onRemoveCity={() => jest.fn()} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
