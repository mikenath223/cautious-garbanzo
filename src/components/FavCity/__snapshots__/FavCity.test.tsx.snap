// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavCityTemp should match snapshot 1`] = `
<div>
  <li>
    <a
      class="card box-shadow"
      href="/details"
      tabindex="0"
    >
      <button
        class="submit-btn remove-btn"
        type="submit"
      >
        -
      </button>
      <div
        class="card-item"
      >
        <img
          alt=""
          class="city-img"
          src="/assets/icons/city.png"
        />
        <h3>
          Lagos
        </h3>
      </div>
      <div
        class="card-item"
      >
        <img
          alt=""
          class="city-img"
          src="/assets/icons/temperature.png"
        />
        <h3>
          25.18
          <sup>
            o
          </sup>
          C
        </h3>
      </div>
    </a>
  </li>
</div>
`;
